src/Features/Core/Portable/EmbeddedLanguages/Json/JsonParser.StrictSyntaxChecker.cs

```diff
@@ -0,0 +1,338 @@
+using Microsoft.CodeAnalysis.EmbeddedLanguages.Common;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.VirtualChars;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.EmbeddedLanguages.Json
+{
+    using static EmbeddedSyntaxHelpers;
+
+    using JsonToken = EmbeddedSyntaxToken<JsonKind>;
+    using JsonTrivia = EmbeddedSyntaxTrivia<JsonKind>;
+
+    internal partial struct JsonParser
+    {
+        /// <summary>
+        /// Checks the superset-tree for constructs that aren't allowed in strict rfc8259
+        /// (https://tools.ietf.org/html/rfc8259) mode.
+        /// </summary>
+        private static class StrictSyntaxChecker
+        {
+            private static EmbeddedDiagnostic? CheckChildren(JsonNode node)
+            {
+                foreach (var child in node)
+                {
+                    var diagnostic = child.IsNode ? CheckSyntax(child.Node) : CheckToken(child.Token);
+                    if (diagnostic != null)
+                    {
+                        return diagnostic;
+                    }
+                }
+
+                return null;
+            }
+
+            private static EmbeddedDiagnostic? CheckToken(JsonToken token)
+                => CheckTrivia(token.LeadingTrivia) ?? CheckTrivia(token.TrailingTrivia);
+
+            private static EmbeddedDiagnostic? CheckTrivia(ImmutableArray<JsonTrivia> triviaList)
+            {
+                foreach (var trivia in triviaList)
+                {
+                    var diagnostic = CheckTrivia(trivia);
+                    if (diagnostic != null)
+                    {
+                        return diagnostic;
+                    }
+                }
+
+                return null;
+            }
+
+            private static EmbeddedDiagnostic? CheckTrivia(JsonTrivia trivia)
+            {
+                switch (trivia.Kind)
+                {
+                    case JsonKind.MultiLineCommentTrivia:
+                    case JsonKind.SingleLineCommentTrivia:
+                        // Strict mode doesn't allow comments at all.
+                        return new EmbeddedDiagnostic(
+                            FeaturesResources.Comments_not_allowed,
+                            GetSpan(trivia.VirtualChars));
+                    case JsonKind.WhitespaceTrivia:
+                        return CheckWhitespace(trivia);
+                }
+
+                return null;
+            }
+
+            private static EmbeddedDiagnostic? CheckWhitespace(JsonTrivia trivia)
+            {
+                foreach (var ch in trivia.VirtualChars)
+                {
+                    switch (ch.Value)
+                    {
+                        case ' ':
+                        case '\t':
+                            break;
+
+                        default:
+                            // Strict mode only allows spaces and horizontal tabs.  Everything else
+                            // is illegal.
```

Comment:
Newlines aren't allowed?

Reply:
Whitespace is different from EndOfLineTrivia